pkg/local_object_storage/blobstor/fstree/head.go

```diff
@@ -0,0 +1,224 @@
+
+		_, err := io.ReadFull(f, comBuf[:])
+		if err != nil {
+			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+				return nil, logicerr.Wrap(apistatus.ErrObjectNotFound)
+			}
+			return nil, err
+		}
+
+		thisOID, l = parseCombinedPrefix(comBuf)
+		if thisOID == nil {
+			return nil, errors.New("malformed combined file")
+		}
+	}
+}
+
+// readObject reads an object from the file.
+func (t *FSTree) readObject(f io.Reader, initial []byte, size int64) (*objectSDK.Object, error) {
+	if size < smallObjectSizeLimit {
+		data, err := t.readFullObject(f, initial, size)
+		if err != nil {
+			return nil, err
+		}
+		var obj objectSDK.Object
+		err = obj.Unmarshal(data)
+		if err != nil {
+			return nil, fmt.Errorf("unmarshal object: %w", err)
+		}
+		return obj.CutPayload(), nil
+	}
+	return t.readUntilPayload(f, initial)
+}
+
+// readFullObject reads full data of object from the file and decompresses it if necessary.
+func (t *FSTree) readFullObject(f io.Reader, initial []byte, size int64) ([]byte, error) {
+	data := make([]byte, size)
+	copy(data, initial)
+	_, err := io.ReadFull(f, data[len(initial):])
+	if err != nil && !errors.Is(err, io.EOF) && !errors.Is(err, io.ErrUnexpectedEOF) {
+		return nil, fmt.Errorf("read: %w", err)
+	}
+
+	return t.Decompress(data)
+}
+
+// readUntilPayload reads an object from the file until the payload field is reached.
+func (t *FSTree) readUntilPayload(f io.Reader, initial []byte) (*objectSDK.Object, error) {
+	data := make([]byte, objectSDK.MaxHeaderLen)
+	copy(data, initial)
+
+	n, err := io.ReadFull(f, data[len(initial):])
+	if err != nil && !errors.Is(err, io.EOF) && !errors.Is(err, io.ErrUnexpectedEOF) {
+		return nil, fmt.Errorf("read: %w", err)
+	}
+	data = data[:len(initial)+n]
+
+	if t.IsCompressed(data) {
+		decoder, err := zstd.NewReader(io.MultiReader(bytes.NewReader(data), f))
+		if err != nil {
+			return nil, fmt.Errorf("zstd decoder: %w", err)
+		}
+		defer decoder.Close()
+
+		buf := make([]byte, objectSDK.MaxHeaderLen)
+		n, err := decoder.Read(buf)
+		if err != nil && !errors.Is(err, io.EOF) {
+			return nil, fmt.Errorf("zstd read: %w", err)
+		}
+		data = buf[:n]
+	}
+
+	return fastExtractHeader(data)
+}
+
+// fastExtractHeader extracts the header of an object from the given byte slice.
+func fastExtractHeader(data []byte) (*objectSDK.Object, error) {
+	var (
+		offset int
+		res    objectSDK.Object
+		obj    object.Object
+	)
+
+	for offset < len(data) {
+		num, typ, n := protowire.ConsumeTag(data[offset:])
+		if n <= 0 {
+			return nil, fmt.Errorf("invalid tag at offset %d", offset)
+		}
+		offset += n
+
+		if typ != protowire.BytesType {
+			return nil, fmt.Errorf("unexpected wire type: %v", typ)
+		}
+
+		if num == fieldObjectPayload {
+			break
+		}
+
+		val, n := protowire.ConsumeBytes(data[offset:])
+		if n <= 0 {
+			return nil, fmt.Errorf("invalid bytes field at offset %d", offset)
+		}
+		offset += n
+
+		switch num {
+		case fieldObjectID:
+			var objID refs.ObjectID
```

Comment:
we could `obj.ObjectId = new(refs.ObjectID)` and get rid of this var. Same for other cases